### ISSUE TITLE
Enum discriminants are not used for the archived type

### ISSUE THREAD

**Issue by djkoloski** [rkyv/rkyv#285]
While enum discriminants can be assigned to the unarchived versions of types, the archived versions cannot be explicitly enumerated in the same way. This means that reordering is a breaking change for the format, which is not great. We should provide a way for users to specify stable enum discriminants for types, either on a per-enum basis or per-variant basis. In the next breaking version, these should just transition to being the discriminants declared in the source.

**Comment by djkoloski** [comment]
If you're using a C-like enum, then you could assign discriminants manually and use `#[archive(as = "Self")]`. You can even specify discriminants for data-carrying enums now that `arbitrary_enum_discriminants` is stable, you'd just handwrite the archived version of your enum and then add `#[archive(as = "Other")]` instead.

**Comment by praveenperera** [comment]
Thanks @djkoloski I'll try `#[archive(as = "Other")]` for new discriminants. There's a lot data already in a database so for the existing ones, if I annotate them as  `#[archive(as = "0")]`,  `#[archive(as = "1")]`, etc would that work?

**Comment by djkoloski** [comment]
Sorry, unfortunately I jumped the gun on this one. I thought that if you assigned explicit enum discriminants to the archived enum, it would cause the serialization code to generate correctly. Unfortunately that's not the case. Here's an example of the problem:

```rust
// Original file
use rkyv::{Archive, Serialize, Deserialize};

#[derive(Archive, Serialize, Deserialize)]
#[archive(as = "ArchivedFoo")]
enum Foo {
    A = 2,
    B = 4,
    C = 6,
}

#[derive(Debug, PartialEq)]
#[repr(C)]
enum ArchivedFoo {
    A = 2,
    B = 4,
    C = 6,
}

fn main() {
    let bytes = rkyv::to_bytes::<_, 0>(&Foo::B).unwrap();
    let archived = unsafe { rkyv::archived_root::<Foo>(&bytes) };
    assert_eq!(archived, &ArchivedFoo::B);
}
```

This is what it expands to:

```rust
use rkyv::{Archive, Serialize, Deserialize};

enum Foo {
    A = 2,
    B = 4,
    C = 6,
}
#[automatically_derived]
///The resolver for an archived [`Foo`]
enum FooResolver {
    ///The resolver for [`Foo::A`]
    #[allow(dead_code)]
    A,
    ///The resolver for [`Foo::B`]
    #[allow(dead_code)]
    B,
    ///The resolver for [`Foo::C`]
    #[allow(dead_code)]
    C,
}
#[automatically_derived]
const _: () = {
    use ::core::marker::PhantomData;
    use ::rkyv::{out_field, Archive, Archived};
    #[repr(u8)]
    enum ArchivedTag {
        A,
        B,
        C,
    }
    impl Archive for Foo {
        type Archived = ArchivedFoo;
        type Resolver = FooResolver;
        #[allow(clippy::unit_arg)]
        #[inline]
        unsafe fn resolve(&self, pos: usize, resolver: Self::Resolver, out: *mut Self::Archived) {
            match resolver {
                FooResolver::A => {
                    out.cast::<ArchivedTag>().write(ArchivedTag::A);
                }
                FooResolver::B => {
                    out.cast::<ArchivedTag>().write(ArchivedTag::B);
                }
                FooResolver::C => {
                    out.cast::<ArchivedTag>().write(ArchivedTag::C);
                }
            }
        }
    }
};
#[automatically_derived]
const _: () = {
    use ::rkyv::{Archive, Fallible, Serialize};
    impl<__S: Fallible + ?Sized> Serialize<__S> for Foo {
        #[inline]
        fn serialize(
            &self,
            serializer: &mut __S,
        ) -> ::core::result::Result<Self::Resolver, __S::Error> {
            Ok(match self {
                Self::A => FooResolver::A,
                Self::B => FooResolver::B,
                Self::C => FooResolver::C,
            })
        }
    }
};
#[automatically_derived]
const _: () = {
    use ::rkyv::{Archive, Archived, Deserialize, Fallible};
    impl<__D: Fallible + ?Sized> Deserialize<Foo, __D> for Archived<Foo> {
        #[inline]
        fn deserialize(&self, deserializer: &mut __D) -> ::core::result::Result<Foo, __D::Error> {
            Ok(match self {
                Self::A => Foo::A,
                Self::B => Foo::B,
                Self::C => Foo::C,
            })
        }
    }
};
#[derive(Debug, PartialEq)]
#[repr(C)]
enum ArchivedFoo {
    A = 2,
    B = 4,
    C = 6,
}

fn main() {
    let bytes = rkyv::to_bytes::<_, 0>(&Foo::B).unwrap();
    let archived = unsafe { rkyv::archived_root::<Foo>(&bytes) };
    assert_eq!(archived, &ArchivedFoo::B);
}
```

The problem here is that the generated `ArchivedTag` doesn't have explicit discriminants, and so the wrong value gets written. Unfortunately, the only way I can see to fix this would be to modify the derive macros to explicitly set the discriminants of that `ArchivedTag`. That's very possible, just a lot of work.

**Comment by praveenperera** [comment]
I see thanks for the update. Maybe for now, a note in the docs would be warranted? We aren't using the C style enums, its Enums with data. 

When we added a new discriminant in the middle of the list, it silently broke the app because some of the saved data was not being deserialized properly. 

A note might be helpful for people using this in the future.

**Comment by djkoloski** [comment]
Fixed by bc5b2f464f5680b651dd8835375779d1c40513f2
